what4/src/What4/Utils/BVDomain.hs

```diff
@@ -1,18 +1,503 @@
+isBVDAny x =
+  case x of
+    BVDAny {} -> True
+    BVDInterval {} -> False
+
+-- | Return unsigned bounds for domain.
+ubounds :: BVDomain w -> (Integer, Integer)
+ubounds a =
+  case a of
+    BVDAny mask -> (0, mask)
+    BVDInterval mask al aw
+      | ah > mask -> (0, mask)
+      | otherwise -> (al, ah)
+      where ah = al + aw
+
+-- | Return signed bounds for domain.
+sbounds :: (1 <= w) => NatRepr w -> BVDomain w -> (Integer, Integer)
+sbounds w a = (lo - delta, hi - delta)
+  where
+    delta = halfRange w
+    (lo, hi) = ubounds (add a (BVDInterval (bvdMask a) delta 0))
+
+-- | Return true if domains contain a common element.
+domainsOverlap :: BVDomain w -> BVDomain w -> Bool
+domainsOverlap a b =
+  case a of
+    BVDAny _ -> True
+    BVDInterval _ al aw ->
+      case b of
+        BVDAny _ -> True
+        BVDInterval mask bl bw ->
+          diff <= bw || diff + aw > mask
+          where diff = (al - bl) .&. mask
+
+eq :: BVDomain w -> BVDomain w -> Maybe Bool
+eq a b
+  | Just x <- asSingleton a
+  , Just y <- asSingleton b = Just (x == y)
+  | domainsOverlap a b == False = Just False
+  | otherwise = Nothing
+
+-- | Check if all elements in one domain are less than all elements in other.
+slt :: (1 <= w) => NatRepr w -> BVDomain w -> BVDomain w -> Maybe Bool
+slt w a b
+  | a_max < b_min = Just True
+  | a_min >= b_max = Just False
+  | otherwise = Nothing
+  where
+    (a_min, a_max) = sbounds w a
+    (b_min, b_max) = sbounds w b
+
+-- | Check if all elements in one domain are less than all elements in other.
+ult :: (1 <= w) => BVDomain w -> BVDomain w -> Maybe Bool
+ult a b
+  | a_max < b_min = Just True
+  | a_min >= b_max = Just False
+  | otherwise = Nothing
+  where
+    (a_min, a_max) = ubounds a
+    (b_min, b_max) = ubounds b
+
+-- | Return @Just@ if every bitvector in the domain has the same bit
+-- at the given index.
+testBit ::
+  NatRepr w ->
+  BVDomain w ->
+  Natural {- ^ Index of bit (least-significant bit has index 0) -} ->
+  Maybe Bool
+testBit w a i =
+  if i >= natValue w then Nothing else
+  case a of
+    BVDAny _ -> Nothing
+    BVDInterval _ al aw
+      | (al `shiftR` j) == (ah `shiftR` j) ->
+        Just (al `Bits.testBit` j)
+      | otherwise ->
+        Nothing
+      where
+        j = fromIntegral i
+        ah = al + aw
+
+--------------------------------------------------------------------------------
+-- Operations
+
+-- | Represents all values
+any :: (1 <= w) => NatRepr w -> BVDomain w
+any w = BVDAny (maxUnsigned w)
+
+-- | Create a bitvector domain representing the integer.
+singleton :: (HasCallStack, 1 <= w) => NatRepr w -> Integer -> BVDomain w
+singleton w x = BVDInterval mask (x .&. mask) 0
+  where mask = maxUnsigned w
+
+-- | @range w l u@ returns domain containing all bitvectors formed
+-- from the @w@ low order bits of some @i@ in @[l,u]@.  Note that per
+-- @testBit@, the least significant bit has index @0@.
+range :: NatRepr w -> Integer -> Integer -> BVDomain w
```

Comment:
Should we `panic` if the range size is negative?

Reply:
Good point. Currently if you do `range w lo hi` with `hi < lo`, you'll get a `BVDomain` value that doesn't satisfy the datatype invariant. I should fix that.

Reply:
Fixed in 39d632544cb7eb3f6b7eaf5526238cc0a8a4c097. Instead of a panic, I just made the function total: `range w a b` will now give you an interval that starts with `a` and continues until you get to `b`. If `b < a` then the interval will simply wrap around.